File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        dummy = ListNode(0, None)
        curr = dummy

        while True:
            # minIdx for picking out the list with smallest node value
            # Note that -1 means nothing has been picked yet
            min_idx = -1

            # Iterate through lists
            for i in range(len(lists)):
                # Continue if lists[i] is empty
                if not lists[i]:
                    continue

                # If we haven't picked anything or
                # if current node's value is smaller than what node we've picked's value,
                # set minIdx to current idx, which means picking that node
                if min_idx == -1 or lists[i].val < lists[min_idx].val:
                    min_idx = i

            # Break out of the while loop if we have nothing to pick
            if min_idx == -1:
                break

            # 1. Attach the picked node to current node
            # 2. Advance current node
            # 3. Advance the list that has been picked for next iteration
            curr.next = lists[min_idx]
            curr = curr.next
            lists[min_idx] = lists[min_idx].next
        return dummy.next

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto mergeTwoLists = [](ListNode* listA, ListNode* listB) -> ListNode* {
            // Create a dummy node that points to nothing
            // and attach a tail node on that dummy node
            ListNode dummy(0, nullptr);
            ListNode* tail = &dummy;

            // Exit the while loop when listA and listB are both exhausted
            while (listA != nullptr && listB != nullptr) {
                // If listA's value is smaller than listB's value
                // 1. Attach listA to tail
                // 2. Advance listA
                if (listA->val < listB->val) {
                    tail->next = listA;
                    listA = listA->next;
                }
                // If listA's value is larger or equal to listB's value
                // 1. Attach listB to tail
                // 2. Advance listB
                else {
                    tail->next = listB;
                    listB = listB->next;
                }

                // Advance tail
                tail = tail->next;
            }

            // Attach the remaining of listA and listB to tail
            tail->next = (listA != nullptr) ? listA : listB;

            // Return the head of the merged list
            return dummy.next;
        };

        // Return nullptr if lists is empty
        if (lists.empty()) {
            return nullptr;
        }

        // Iterate through lists from index 1 and merge them one by one
        for (int i = 1; i < static_cast<int>(lists.size()); ++i) {
            lists[i] = mergeTwoLists(lists[i], lists[i - 1]);
        }
        return lists.back();
    }
};